### PR TITLE
feat(provider/gce): Support named ports for global LBs.

### PR DESCRIPTION
**Pull request by jtk54**
Added a new section to declare named ports in the affected LBs and a new section in the GCE server group wizard to add named ports and map them to an actual port number.

### CODE REVIEW

**Comment by jtk54** [comment]
Depends on https://github.com/spinnaker/clouddriver/pull/1930.

**Comment by jtk54** [comment]
Ping.

File: app/scripts/modules/google/src/loadBalancer/configure/http/backendService/backendService.component.html [hunk 1]
```diff
@@ -103,6 +103,19 @@
 
     <div class="form-group">
       <div class="col-md-4 sm-label-right">
+        <b>Port Name</b>
+        <help-field key="gce.loadBalancer.portName" />
```
Comment (danielpeach) [inline app/scripts/modules/google/src/loadBalancer/configure/http/backendService/backendService.component.html]: Needs a closing tag: `<help-field ... ></help-field>`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.html [hunk 2]
```diff
@@ -1,19 +1,56 @@
...
+            <ui-select ng-if="!namedPort.name || !namedPort.name.length"
+                       ng-model="namedPort.name"
+                       class="form-control input-sm">
+              <ui-select-match>{{$item.name}}</ui-select-match>
+              <ui-select-choices repeat="portName in $ctrl.getPortNames()">
+                <span ng-bind-html="portName"></span>
+              </ui-select-choices>
+            </ui-select>
+
+            <input ng-if="namedPort.name.length"
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.html]: Why does this need to become a read-only input field after you pick an option? Can't you just pick again from the select field?

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
-import { set, get, has, without } from 'lodash';
+import { set, get, has, without, intersection, uniq } from 'lodash';
 import './loadBalancingPolicySelector.component.less';
 
 class GceLoadBalancingPolicySelectorController implements IController {
 
   public maxPort = 65535;
   public command: any;
   [key: string]: any;
+  public globalBackendServices: any[];
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: `IGceBackendService[]`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 4]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
     return balancingModes;
   }
 
-  public $onDestroy (): void {
+  public $onInit(): void {
+    this.gceBackendServiceReader.listBackendServices('globalBackendService')
+      .then((services: any[]) => {
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: `IGceBackendServices[]`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 5]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
     delete this.command.loadBalancingPolicy;
   }
 
-  private decimalToPercent (value: number): number {
+  public addNamedPort() {
+    if (!this.command.loadBalancingPolicy.namedPorts) {
+      this.command.loadBalancingPolicy.namedPorts = [];
+    }
+
+    this.command.loadBalancingPolicy.namedPorts.push({'name': '', 'port': 80});
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: Omit the quotes from the object keys.

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.html [hunk 6]
```diff
@@ -1,19 +1,56 @@
...
+                   name="listeningPort"
+                   required
+                   class="form-control input-sm"
+                   ng-model="namedPort.port"
+                   max="{{ $ctrl.maxPort }}"
+                   min="1"/>
+          </div>
+          <div class="col-md-2 sm-label-left">
+            <span class="glyphicon glyphicon-trash" uib-tooltip="Remove Named Port"
+                  ng-click="$ctrl.removeNamedPort(namedPort)"></span>
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.html]: Inside an `ng-repeat`, you can reference `$index`. If you pass `$index` to `$ctrl.removeNamedPort`, you can rewrite `$ctrl.removeNamedPort`:

```
public removeNamedPort(index: number): void {
    this.command.loadBalancingPolicy.namedPorts.splice(index, 1);
}
```

This also helps you avoid some nasty problems that would show up if the same named port was listed twice (which  probably shouldn't happen, and ideally would be validated).

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 7]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
+    const index = this.command.loadBalancingPolicy.namedPorts.indexOf(namedPort);
+    this.command.loadBalancingPolicy.namedPorts.splice(index, 1);
+  }
+
+  public getPortNames(): string[] {
+    const index = this.command.backingData.filtered.loadBalancerIndex;
+    const selected = this.command.loadBalancers;
+    const inUsePortNames = this.command.loadBalancingPolicy.namedPorts.map(namedPort => namedPort.name);
+
+    const getThem = (globalBackendServices: any[], loadBalancer: any): string[] => {
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: `(globalBackendServices: IGceBackendServices[], loadbalancer: string): string[] => { ... }`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 8]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
+      switch (get(index[loadBalancer], 'loadBalancerType')) {
+        case 'SSL':
+        case 'TCP':
+        case 'HTTP':
+          const lbBackendServices = get(index[loadBalancer], 'backendServices');
+          const filteredBackendServices = globalBackendServices.filter((service: any) => lbBackendServices.indexOf(service.name) > -1);
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: Switch `.indexOf(service.name) > -1` with `.includes(service.name)`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 9]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
+          const portNames =  filteredBackendServices.map((service: any) => service.portName);
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: `IGceBackendService`

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 10]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
+          const portNameIntersection = intersection(portNames, inUsePortNames);
+          return portNames.filter(portName =>  portNameIntersection.indexOf(portName) === -1);
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: Switch `indexOf` with `includes`.

File: app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts [hunk 11]
```diff
@@ -67,11 +72,54 @@ class GceLoadBalancingPolicySelectorController implements IController {
...
+        default:
+          return [];
+      }
+    };
+
+    return uniq(selected.map(lbName => getThem(this.globalBackendServices, lbName)).reduce((l: string[], r: string[]) => l.concat(r)));
```
Comment (danielpeach) [inline app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts]: I think you're looking for `flatMap` here:

```
import { chain } from 'lodash';

...

return chain(selected).flatMap(lbName => getThem(...)).uniq().value();
```

**Pull request merged**